### ISSUE TITLE
Out of memory error on build box

### ISSUE THREAD

**Issue by agoodno** [sbt/sbt#101]
I've got a large, java-only project that I'm building with sbt (all of our new development is in Scala and we'd like to standardize on one build tool).  One project has 3200 source files in it and another that depends on it has 2700 source files.  It builds fine on my Linux workstation as is but runs out of memory on our Jenkins box.  I've tried turning off parallel execution and it seemed to help because it compiled the first project but still runs out of mem on the second.  The only difference I know of is that Jenkins is spawning the process and the JVM on that box is a 64-bit sun jvm instead of my local 32-bit jvm.  I've eliminated Jenkins by firing off the build from the command-line - same error.  This multi-module project compiles fine under ant, gradle, and sbt 0.7.7.

Here's information dump from attaching with jmap while the process is hanging:

Started by user anonymous
Updating <svn_server_root>/cc/trunk
Fetching '<svn_server_root>/CC/trunk/CC-Conversion/src' at -1 into '/var/data/hudson/jobs/cc_trunk/workspace/cc_trunk/cc-conversion/src/main/java'
At revision 231844
Fetching '<svn_server_root>/CC/trunk/CC-Form/src' at -1 into '/var/data/hudson/jobs/cc_trunk/workspace/cc_trunk/cc-form/src/main/resources'
At revision 231844
Fetching '<svn_server_root>/CC/trunk/CC-Server/src' at -1 into '/var/data/hudson/jobs/cc_trunk/workspace/cc_trunk/cc-server/src/main/java'
At revision 231844
Fetching '<svn_server_root>/CC/trunk/Calendar-Kiosk/src' at -1 into '/var/data/hudson/jobs/cc_trunk/workspace/cc_trunk/calendar-kiosk/src/main/java'
At revision 231844
Fetching '<svn_server_root>/CC/trunk/Step-CC/src' at -1 into '/var/data/hudson/jobs/cc_trunk/workspace/cc_trunk/cc-step/src/main/java'
At revision 231844
Fetching '<svn_server_root>/CC/trunk/CC-Client/src' at -1 into '/var/data/hudson/jobs/cc_trunk/workspace/cc_trunk/cc-client/src/main/java'
At revision 231844
Fetching '<svn_server_root>/CC/trunk/CC-Client-Server/src' at -1 into '/var/data/hudson/jobs/cc_trunk/workspace/cc_trunk/cc-client-server/src/main/java'
At revision 231844
Fetching '<svn_server_root>/CC/trunk/CC-Interfaces/src' at -1 into '/var/data/hudson/jobs/cc_trunk/workspace/cc_trunk/cc-interfaces/src/main/java'
At revision 231844
Fetching '<svn_server_root>/CC/trunk/CC-XSL/src' at -1 into '/var/data/hudson/jobs/cc_trunk/workspace/cc_trunk/cc-xsl/src/main/resources'
At revision 231844
At revision 231844
no change for <svn_server_root>/cc/trunk since the previous build
no change for <svn_server_root>/CC/trunk/CC-Conversion/src since the previous build
no change for <svn_server_root>/CC/trunk/CC-Form/src since the previous build
no change for <svn_server_root>/CC/trunk/CC-Server/src since the previous build
no change for <svn_server_root>/CC/trunk/Calendar-Kiosk/src since the previous build
no change for <svn_server_root>/CC/trunk/Step-CC/src since the previous build
no change for <svn_server_root>/CC/trunk/CC-Client/src since the previous build
no change for <svn_server_root>/CC/trunk/CC-Client-Server/src since the previous build
no change for <svn_server_root>/CC/trunk/CC-Interfaces/src since the previous build
no change for <svn_server_root>/CC/trunk/CC-XSL/src since the previous build
[cc_trunk] $ /bin/java -Xss4m -Xms256m -Xmx1536m -XX:PermSize=128m -XX:MaxPermSize=256m -XX:+CMSClassUnloadingEnabled -noverify -XX:-UseGCOverheadLimit -Dsbt.log.noformat=true -jar /opt/sbt/sbt-launch-0.10.0.jar clean update compile
[info] Set current project to root (in build file:/var/data/hudson/jobs/cc_trunk/workspace/cc_trunk/)
[success] Total time: 12 s, completed Jul 12, 2011 8:16:23 AM
[info] Updating...
[info] Done updating.
[info] Updating...
[info] Done updating.
[info] Updating...
[info] Done updating.
[info] Updating...
[info] Done updating.
[info] Updating...
[info] Done updating.
[info] Updating...
[info] Done updating.
[info] Updating...
[info] Done updating.
[info] Updating...
[info] Done updating.
[info] Updating...
[info] Done updating.
[info] Updating...
[info] Done updating.
[success] Total time: 17 s, completed Jul 12, 2011 8:16:40 AM
[info] Updating...
[info] Done updating.
[info] Compiling 3225 Java sources to /var/data/hudson/jobs/cc_trunk/workspace/cc_trunk/cc-client-server/target/scala-2.9.0.1/classes...
[error] Note: Some input files use unchecked or unsafe operations.
[error] Note: Recompile with -Xlint:unchecked for details.
[info] Compiling 2797 Java sources to /var/data/hudson/jobs/cc_trunk/workspace/cc_trunk/cc-client/target/scala-2.9.0.1/classes...
[error] Note: Some input files use or override a deprecated API.
[error] Note: Recompile with -Xlint:deprecation for details.
[error] Note: Some input files use unchecked or unsafe operations.
[error] Note: Recompile with -Xlint:unchecked for details.
2011-07-12 08:25:27
Full thread dump Java HotSpot(TM) 64-Bit Server VM (11.3-b02 mixed mode):

"pool-3-thread-1" prio=10 tid=0x00002aab21e10400 nid=0x1241 runnable [0x0000000042539000..0x000000004253baa0]
   java.lang.Thread.State: RUNNABLE
    at scala.reflect.ClassManifest$.classType(ClassManifest.scala:181)
    at sbt.ClassToAPI$.pathFromString(ClassToAPI.scala:195)
    at sbt.ClassToAPI$.reference(ClassToAPI.scala:175)
    at sbt.ClassToAPI$.reference(ClassToAPI.scala:164)
    at sbt.ClassToAPI$.reference(ClassToAPI.scala:191)
    at sbt.ClassToAPI$.referenceP(ClassToAPI.scala:181)
    at sbt.ClassToAPI$.reference(ClassToAPI.scala:189)
    at sbt.ClassToAPI$.parameter(ClassToAPI.scala:97)
    at sbt.ClassToAPI$$anonfun$22.apply(ClassToAPI.scala:87)
    at sbt.ClassToAPI$$anonfun$22.apply(ClassToAPI.scala:87)
    at scala.Tuple3$Zipped$$anonfun$map$1.apply(Tuple3.scala:66)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:34)
    at scala.collection.mutable.ArrayOps.foreach(ArrayOps.scala:35)
    at scala.Tuple3$Zipped.map(Tuple3.scala:64)
    at sbt.ClassToAPI$.defLike(ClassToAPI.scala:87)
    at sbt.ClassToAPI$.methodToDef(ClassToAPI.scala:78)
    at sbt.ClassToAPI$$anonfun$7.apply(ClassToAPI.scala:47)
    at sbt.ClassToAPI$$anonfun$7.apply(ClassToAPI.scala:47)
    at sbt.ClassToAPI$$anonfun$mergeMap$1.apply(ClassToAPI.scala:107)
    at sbt.ClassToAPI$$anonfun$mergeMap$1.apply(ClassToAPI.scala:107)
    at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:227)
    at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:227)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:34)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:32)
    at scala.collection.TraversableLike$class.flatMap(TraversableLike.scala:227)
    at scala.collection.mutable.WrappedArray.flatMap(WrappedArray.scala:32)
    at sbt.ClassToAPI$.merge(ClassToAPI.scala:113)
    at sbt.ClassToAPI$.mergeMap(ClassToAPI.scala:107)
    at sbt.ClassToAPI$.structure(ClassToAPI.scala:47)
    at sbt.ClassToAPI$.x$1$1(ClassToAPI.scala:37)
    - locked <0x00002aaabf5082f0> (a sbt.ClassToAPI$)
    at sbt.ClassToAPI$.instance$1(ClassToAPI.scala:37)
    - locked <0x00002aaabf5082f0> (a sbt.ClassToAPI$)
    at sbt.ClassToAPI$$anonfun$5.apply(ClassToAPI.scala:38)
    at sbt.ClassToAPI$$anonfun$5.apply(ClassToAPI.scala:38)
    at xsbti.SafeLazy$Impl._t(SafeLazy.scala:20)
    - locked <0x00002aaafbdf0da8> (a xsbti.SafeLazy$Impl)
    at xsbti.SafeLazy$Impl.get(SafeLazy.scala:24)
    at xsbti.api.ClassLike.structure(ClassLike.java:27)
    at xsbt.api.TagTypeVariables$$anonfun$tagClass0$1.apply$mcV$sp(TagTypeVariables.scala:45)
    at xsbt.api.TagTypeVariables$$anonfun$tagParameterizedDefinition$1.apply$mcV$sp(TagTypeVariables.scala:66)
    at xsbt.api.TagTypeVariables.scope(TagTypeVariables.scala:140)
    at xsbt.api.TagTypeVariables.tagParameterizedDefinition(TagTypeVariables.scala:64)
    at xsbt.api.TagTypeVariables.tagClass0(TagTypeVariables.scala:43)
    at xsbt.api.TagTypeVariables.tagClass(TagTypeVariables.scala:41)
    at xsbt.api.TagTypeVariables.tagDefinition(TagTypeVariables.scala:34)
    at xsbt.api.TagTypeVariables$$anonfun$tag$1.apply(TagTypeVariables.scala:26)
    at xsbt.api.TagTypeVariables$$anonfun$tag$1.apply(TagTypeVariables.scala:26)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:34)
    at scala.collection.mutable.ArrayOps.foreach(ArrayOps.scala:35)
    at xsbt.api.TagTypeVariables.tag(TagTypeVariables.scala:26)
    at xsbt.api.TagTypeVariables$.apply(TagTypeVariables.scala:12)
    at xsbt.api.SameAPI$.apply(SameAPI.scala:58)
    at sbt.inc.Incremental$.sameSource(Incremental.scala:62)
    at sbt.inc.Incremental$$anonfun$3.apply(Incremental.scala:54)
    at sbt.inc.Incremental$$anonfun$3.apply(Incremental.scala:54)
    at scala.Tuple3$Zipped$$anonfun$filter$1.apply(Tuple3.scala:97)
    at scala.collection.immutable.HashSet$HashSet1.foreach(HashSet.scala:123)
    at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:322)
    at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:322)
    at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:322)
    at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:322)
    at scala.Tuple3$Zipped.filter(Tuple3.scala:93)
    at sbt.inc.Incremental$.changedIncremental(Incremental.scala:54)
    at sbt.inc.Incremental$.cycle(Incremental.scala:37)
    at sbt.inc.Incremental$.compile(Incremental.scala:20)
    at sbt.inc.IncrementalCompile$.apply(Compile.scala:17)
    at sbt.AggressiveCompile.compile2(AggressiveCompile.scala:86)
    at sbt.AggressiveCompile.compile1(AggressiveCompile.scala:37)
    at sbt.AggressiveCompile.apply(AggressiveCompile.scala:27)
    at sbt.Compiler$.apply(Compiler.scala:98)
    at sbt.Defaults$$anonfun$compileTask$1.apply(Defaults.scala:408)
    at sbt.Defaults$$anonfun$compileTask$1.apply(Defaults.scala:408)
    at sbt.Scoped$$anonfun$hf2$1.apply(Structure.scala:507)
    at sbt.Scoped$$anonfun$hf2$1.apply(Structure.scala:507)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:44)
    at sbt.Scoped$Reduced$$anonfun$combine$1$$anonfun$apply$9.apply(Structure.scala:305)
    at sbt.Scoped$Reduced$$anonfun$combine$1$$anonfun$apply$9.apply(Structure.scala:305)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:39)
    at sbt.std.Transform$$anon$5.work(System.scala:66)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:220)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:220)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:13)
    at sbt.Execute.work(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:220)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:220)
    at sbt.CompletionService$$anon$1$$anon$2.call(CompletionService.scala:26)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
    at java.util.concurrent.FutureTask.run(FutureTask.java:138)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
    at java.util.concurrent.FutureTask.run(FutureTask.java:138)
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
    at java.lang.Thread.run(Thread.java:619)

"Low Memory Detector" daemon prio=10 tid=0x00002aab1ec24400 nid=0x1234 runnable [0x0000000000000000..0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"CompilerThread1" daemon prio=10 tid=0x00002aab1ec21800 nid=0x1233 waiting on condition [0x0000000000000000..0x0000000041c37330]
   java.lang.Thread.State: RUNNABLE

"CompilerThread0" daemon prio=10 tid=0x00002aab1ec1d800 nid=0x1232 waiting on condition [0x0000000000000000..0x0000000041b363b0]
   java.lang.Thread.State: RUNNABLE

"Signal Dispatcher" daemon prio=10 tid=0x00002aab1ec1bc00 nid=0x1231 waiting on condition [0x0000000000000000..0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" daemon prio=10 tid=0x00002aab1e973c00 nid=0x1230 in Object.wait() [0x0000000041635000..0x0000000041635e20]
   java.lang.Thread.State: WAITING (on object monitor)
    at java.lang.Object.wait(Native Method)
    - waiting on <0x00002aaabe839910> (a java.lang.ref.ReferenceQueue$Lock)
    at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:116)
    - locked <0x00002aaabe839910> (a java.lang.ref.ReferenceQueue$Lock)
    at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:132)
    at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:159)

"Reference Handler" daemon prio=10 tid=0x00002aab1e972000 nid=0x122f in Object.wait() [0x0000000041234000..0x0000000041234ca0]
   java.lang.Thread.State: BLOCKED (on object monitor)
    at java.lang.Object.wait(Native Method)
    - waiting on <0x00002aaabe89fac0> (a java.lang.ref.Reference$Lock)
    at java.lang.Object.wait(Object.java:485)
    at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:116)
    - locked <0x00002aaabe89fac0> (a java.lang.ref.Reference$Lock)

"main" prio=10 tid=0x0000000040110c00 nid=0x1225 waiting on condition [0x0000000040529000..0x000000004052af60]
   java.lang.Thread.State: WAITING (parking)
    at sun.misc.Unsafe.park(Native Method)
    - parking to wait for  <0x00002aaabeb03f78> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
    at java.util.concurrent.locks.LockSupport.park(LockSupport.java:158)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:1925)
    at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:358)
    at java.util.concurrent.ExecutorCompletionService.take(ExecutorCompletionService.java:164)
    at sbt.CompletionService$$anon$1.take(CompletionService.scala:29)
    at sbt.Execute.next$1(Execute.scala:74)
    at sbt.Execute.processAll(Execute.scala:77)
    at sbt.Execute.runKeep(Execute.scala:57)
    at sbt.Execute.run(Execute.scala:51)
    at sbt.EvaluateTask$.runTask(EvaluteTask.scala:83)
    at sbt.Aggregation$$anonfun$7.apply(Aggregation.scala:85)
    at sbt.Aggregation$$anonfun$7.apply(Aggregation.scala:85)
    at sbt.EvaluateTask$.withStreams(EvaluteTask.scala:65)
    at sbt.Aggregation$.runTasks(Aggregation.scala:85)
    at sbt.Aggregation$$anonfun$applyTasks$1.apply(Aggregation.scala:74)
    at sbt.Aggregation$$anonfun$applyTasks$1.apply(Aggregation.scala:73)
    at sbt.Command$$anonfun$applyEffect$2$$anonfun$apply$3.apply(Command.scala:63)
    at sbt.Command$$anonfun$applyEffect$2$$anonfun$apply$3.apply(Command.scala:63)
    at sbt.Command$.process(Command.scala:91)
    at sbt.MainLoop$$anonfun$next$1$$anonfun$apply$1.apply(Main.scala:72)
    at sbt.MainLoop$$anonfun$next$1$$anonfun$apply$1.apply(Main.scala:72)
    at sbt.State$$anon$1.process(State.scala:54)
    at sbt.MainLoop$$anonfun$next$1.apply(Main.scala:72)
    at sbt.MainLoop$$anonfun$next$1.apply(Main.scala:72)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:13)
    at sbt.MainLoop$.next(Main.scala:72)
    at sbt.MainLoop$.run(Main.scala:63)
    at sbt.xMain.run(Main.scala:33)
    at xsbt.boot.Launch$.run(Launch.scala:54)
    at xsbt.boot.Launch$$anonfun$explicit$1.apply(Launch.scala:43)
    at xsbt.boot.Launch$.launch(Launch.scala:68)
    at xsbt.boot.Launch$.apply(Launch.scala:14)
    at xsbt.boot.Boot$.runImpl(Boot.scala:24)
    at xsbt.boot.Boot$.main(Boot.scala:15)
    at xsbt.boot.Boot.main(Boot.scala)

"VM Thread" prio=10 tid=0x00002aab1e96c800 nid=0x122e runnable 

"GC task thread#0 (ParallelGC)" prio=10 tid=0x000000004011bc00 nid=0x1226 runnable 

"GC task thread#1 (ParallelGC)" prio=10 tid=0x000000004011d400 nid=0x1227 runnable 

"GC task thread#2 (ParallelGC)" prio=10 tid=0x000000004011f000 nid=0x1228 runnable 

"GC task thread#3 (ParallelGC)" prio=10 tid=0x0000000040120800 nid=0x1229 runnable 

"GC task thread#4 (ParallelGC)" prio=10 tid=0x0000000040122000 nid=0x122a runnable 

"GC task thread#5 (ParallelGC)" prio=10 tid=0x0000000040123c00 nid=0x122b runnable 

"GC task thread#6 (ParallelGC)" prio=10 tid=0x0000000040125400 nid=0x122c runnable 

"GC task thread#7 (ParallelGC)" prio=10 tid=0x0000000040127000 nid=0x122d runnable 

"VM Periodic Task Thread" prio=10 tid=0x00002aab1ec26400 nid=0x1235 waiting on condition 

JNI global references: 1186

Heap
 PSYoungGen      total 349568K, used 174211K [0x00002aaafe260000, 0x00002aab1e260000, 0x00002aab1e260000)
  eden space 174848K, 99% used [0x00002aaafe260000,0x00002aab08c80fd8,0x00002aab08d20000)
  from space 174720K, 0% used [0x00002aab08d20000,0x00002aab08d20000,0x00002aab137c0000)
  to   space 174720K, 0% used [0x00002aab137c0000,0x00002aab137c0000,0x00002aab1e260000)
 PSOldGen        total 1048576K, used 1048575K [0x00002aaabe260000, 0x00002aaafe260000, 0x00002aaafe260000)
  object space 1048576K, 99% used [0x00002aaabe260000,0x00002aaafe25ffd8,0x00002aaafe260000)
 PSPermGen       total 131328K, used 94205K [0x00002aaaae260000, 0x00002aaab62a0000, 0x00002aaabe260000)
  object space 131328K, 71% used [0x00002aaaae260000,0x00002aaab3e5f7d8,0x00002aaab62a0000)

FATAL: command execution failed: null
java.lang.InterruptedException
    at java.lang.Object.wait(Native Method)
    at java.lang.Object.wait(Object.java:485)
    at java.lang.UNIXProcess.waitFor(UNIXProcess.java:165)
    at hudson.Proc$LocalProc.join(Proc.java:319)
    at hudson.Launcher$ProcStarter.join(Launcher.java:336)
    at org.jvnet.hudson.plugins.SbtPluginBuilder.perform(SbtPluginBuilder.java:108)
    at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:19)
    at hudson.model.AbstractBuild$AbstractRunner.perform(AbstractBuild.java:664)
    at hudson.model.Build$RunnerImpl.build(Build.java:177)
    at hudson.model.Build$RunnerImpl.doRun(Build.java:139)
    at hudson.model.AbstractBuild$AbstractRunner.run(AbstractBuild.java:430)
    at hudson.model.Run.run(Run.java:1376)
    at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:46)
    at hudson.model.ResourceController.execute(ResourceController.java:88)
    at hudson.model.Executor.run(Executor.java:175)
Build step 'Build using sbt' changed build result to FAILURE
Build step 'Build using sbt' marked build as failure
Sending e-mails to: andrew.goodnough@wicourts.gov
Finished: FAILURE

Thanks,

Andy

**Comment by agoodno** [comment]
This was fixed by increasing the upper limit on the heap to 4096 using -Xmx4096m.

**Comment by harrah** [comment]
This isn't entirely without merit, but it is probably a duplicate of #21.
